### DISCUSSION TITLE
Why does the Nvidia driver include confidential computing-related code for architectures other than Hopper?

### DISCUSSION THREAD

**Discussion by VicoinDub** [NVIDIA/open-gpu-kernel-modules#718] [Q&A]
From what I know, the concept of confidential computing was introduced with the Hopper architecture.
However, I’ve noticed that in the Nvidia driver, there’s confidential computing-related code for other architectures like Pascal and Ampere as well.
I’m a bit puzzled by this—does anyone happen to know why?

For example, there are many cases like this one:
https://github.com/NVIDIA/open-gpu-kernel-modules/blob/550.90.07/kernel-open/nvidia-uvm/uvm_pascal_fault_buffer.c#L287-L302
Thanks!

**Comment by steven-bellock** [comment]
HALs are wired so that software that supports newer GPUs can use the code for older GPUs, provided that makes sense. If it doesn't make sense then a new function can be implemented. In your example, `g_uvm_global.conf_computing_enabled` would be always be `false` for GPUs older than Hopper.
